src/types/multiroot.rs

```diff
@@ -0,0 +1,286 @@
+    }
+
+    /// This function returns the current estimate of the root for the solver `s`, given by `s->x`.
+    pub fn root(& self) -> ::VectorF64 {
+        ::VectorF64::soft_wrap(unsafe {sys::gsl_multiroot_fsolver_root(self.unwrap_shared())})
+    }
+
+    /// This function returns the last step `dx` taken by the solver `s`, given by `s->dx`.
+    pub fn dx(& self) -> ::VectorF64 {
+        ::VectorF64::soft_wrap(unsafe {sys::gsl_multiroot_fsolver_dx(self.unwrap_shared())})
+    }
+
+    /// This function returns the function value `f(x)` at the current estimate of the root for
+    /// the solver `s`, given by `s->f`.
+    pub fn f(& self) -> ::VectorF64 {
+        ::VectorF64::soft_wrap(unsafe {sys::gsl_multiroot_fsolver_f(self.unwrap_shared())})
+    }
+
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use VectorF64;
+    use multiroot::test_residual;
+
+    /// checking a test function
+    /// must return a success criteria (or failure)
+    fn rosenbrock_f(x: & ::VectorF64, f: &mut ::VectorF64) -> ::Value{
+        f.set(0, 1.0 - x.get(0));
+        f.set(1, x.get(0) - x.get(1).powf(2.0));
+        ::Value::Success
+    }
+
+    fn print_state(solver: &mut MultiRootFSolver, iteration: usize) {
+        let f = solver.f();
+        let x = solver.root();
+        println!("iter: {}, f = [{:+.2e}, {:+.2e}], x = [{:+.5}, {:+.5}]",
+                 iteration,
+                 f.get(0),
+                 f.get(1),
+                 x.get(0),
+                 x.get(1))
+    }
+
+    #[test]
+    fn test_multiroot_fsolver() {
+        // setup workspace
+        let mut multi_root = MultiRootFSolver::new(&MultiRootFSolverType::hybrid(), 2).unwrap();
+        let array_size: usize = 2;
+        let guess_value = VectorF64::from_slice(&[-10.0, -5.0]).unwrap();
+        multi_root.set(&rosenbrock_f, array_size,&guess_value);
+
+        // iteration counters
+        let max_iter: usize = 100;
+        let mut iter = 0;
+
+        // convergence checks
+        let mut status = ::Value::Continue;
+        let epsabs = 1e-6;
+
+        print_state(&mut multi_root, 0);
+
+        while matches!(status, ::Value::Continue) && iter < max_iter {
+            // iterate solver
+            status = multi_root.iterate();
+
+            // check iteration for failure (e.g. BadEquation (f set to inf or NaN) or
+            // ENoProg (not or bad progress).
+            if !matches!(status, ::Value::Success) {
+                break;
+            }
+
+            // print current iteration
+            print_state(&mut multi_root, iter);
+
+            // test for convergence
+            let f_value = multi_root.f();
+            status = test_residual(&f_value, epsabs);
+
+            // check if iteration succeeded
+            if matches!(status, ::Value::Success) {
+                println!("Converged");
+            }
+
+            iter += 1;
+        }
+        assert!(matches!(status, ::Value::Success))
+    }
+}
```

Comment:
Please add a backline at the end of the file.

Reply:
added